### ISSUE TITLE
ScrollSpy not behaving correctly

### ISSUE THREAD

**Issue by omar-abdul** [twbs/bootstrap#36431]
Scroll spy on body element will only activate the first item on the nav menu and will not update accordingly, Boostrap 5.1.3 works fine but for some reason 5.2 beta doesn't

**Comment by mdo** [comment]
Bug reports must include a **live demo** of the issue. Per our [contributing guidelines](https://github.com/twbs/bootstrap/blob/master/.github/CONTRIBUTING.md), please create a reduced test case via [CodePen](https://codepen.io/) or [JS Bin](https://jsbin.com/) and report back with your link, Bootstrap version, and specific browser and operating system details.

---

_This is a saved reply._

**Comment by mastercoms** [comment]
Scrollspy has clearly regressed, even on the docs. The prior behavior was to make a link active if its associated element reached the scroll top. Now, it seems to clear the active state if one element is more visible than the other. This does not get the same results, and has very strange behavior in various different scenarios which expected the previous behavior. This is just the coded behavior, it's not browser dependent.

https://user-images.githubusercontent.com/2672245/170849218-7faab47f-67e6-4295-8265-8544725b4d35.mp4

.

**Comment by mastercoms** [comment]
Here's one example of how it broke.

https://user-images.githubusercontent.com/2672245/170849451-0552fe36-70dd-4311-96e9-8de918a9631f.mp4

There's just no active link, since the heading is not visible. There is absolutely no way to handle this correctly, since if you have a wrapper element around each heading and paragraph, then it does not activate properly at all (since the large wrapper element will not satisfy the thresholds properly). And you can fiddle around with the root margin and maybe thresholds to help, but really it will be specific to each section, so there is no setup that really works with varying bits of spied content sections.

**Comment by tranqt1984** [comment]
I have the same issue with version 5.2 regarding the ScrollSpy skipping to the newly observed element when scrolling. This does not provide a good user experience and will confuse the user as they are scrolling. 

In this video, I updated the height from 200px to 600px which shows the same issue I'm having when using the ScrollSpy. 

https://user-images.githubusercontent.com/62732616/173135284-76a83643-1b23-4448-b2c4-1eadb4db0d20.mov

POC: Targeting the top element when scrolling. https://codepen.io/byondsick12/pen/abqjYzg
POC:  IntersectionObserver observing the element that is being scrolled into the viewport which mimics the ScrollSpy buggy behavior. https://codepen.io/byondsick12/pen/oNEdqzW

**Comment by julien-deramond** [comment]
Reopened it since a CodePen has been provided by @tranqt1984 in the meantime (haven't checked its content yet)

**Comment by GeoSot** [comment]
@kenrobbins #36750 this may help you to tweak it properly

**Comment by Khaldei** [comment]
There are additional pieces that are not working here.  They are visible directly on the demo site.  For example, navigate to...

[https://getbootstrap.com/docs/5.0/components/scrollspy/#item-1-2](url)

Scroll to "Example with nested nav"

You will notice that the scrollspy navbar is now off by two elements.

"Item 1" in the navbar is now linked to "Item 1-2"
![image](https://user-images.githubusercontent.com/23502821/182466194-c4af2ab2-4414-4948-ad6f-05883fb2a8b8.png)

If you scroll up to the "Item 1", nothing at all is highlighted.
![image](https://user-images.githubusercontent.com/23502821/182466343-a5e586f8-7e85-46ef-ad60-6cef32b138a3.png)

And if you scroll down, the offsets continue.
![image](https://user-images.githubusercontent.com/23502821/182466497-ae44c28e-56a0-441d-a2c2-46fc5be46e54.png)

**Comment by qofe** [comment]
Posting in case it helps anyone else -- As per the original post about Scrollspy not working on the 'body' element in 5.2. I was going through the steps to make a code-pen and found through that exercise it worked just fine. My issue was a conflict I had in the 'overflow' attribute on the 'body'. Once I cleared that up, it worked as expected.

**Comment by shikkaba** [comment]
>  My issue was a conflict I had in the 'overflow' attribute on the 'body'. Once I cleared that up, it worked as expected.

@qofe Can you elaborate on this?

**Comment by iMattPro** [comment]
Is there any update on this? Scrollspy and the Docs are awful now. Before Scrollspy when spying on a list of links, would highlight when the target reached the top of the viewport, but now it highlights just when the target is visible in the window anywhere. There's absolutely no guidance on how to tweak these root margins and thresholds to at least recreate the original scroll spy behavior from before 5.2

**Comment by usa-usa-usa-usa** [comment]
I am experiencing the same issue. If a section is too short to consume enough of the vieweport, the nav item is never actually navigating.

I'd prefer that whatever is at the top of the screen is what is deemed "active".

**Comment by SekarMurugesan** [comment]
same issue i have in react when moving downwards it working but moving upward it is not working

**Comment by hansceelen** [comment]
When even the examples on the Bootstrap 5.2 docs don't work as expected (as noted by @Khaldei and @tranqt1984), I just state that scrollspy has become unusable the way it is right now and should be rolled back. Scrollspy should simply be an active state on a defined y-position, until the next target is triggered and vice versa. I.e.... the way it was. Nice and simple. For my current project i'll have to drop 5.2 scrollspy and write one myself.

**Comment by marcus-at-localhost** [comment]
@shikkaba I experienced the same as @qofe.
The issue, described in the very first post, is having something like this:

```css
body{
	overflow-x: hidden; /* or auto or scroll */
}
```

and it was resolved by doing something like this:

```css
body{
	overflow-x: unset; /* or initial */
}
```

I don't say that's generally the case for this not working, but overflow is likely the issue.

**Comment by leroynas** [comment]
I am actually working on a custom scroll-spy for a company. I was trying to get inspiration by checking the bootstrap implementation, however I found out that the `IntersectionObserver` only returns entries that **changed** in visibility. However it can be that there are 10 entries visible at load. These entries are processed, and the first entry is visible (almost always during load). When you scroll down, only entries that change in visibility are changed so 11, then 12, then 13 (with `entry.isIntersection` true). And likewise elements 1, then 2, then 3 (with `entry.isIntersection` false).

On page load
```
|--------------------------------------------------|
|  item 1                                          |
|  item 2                                          |
|  item 3                                          |
|  item 4                                          |
|--------------------------------------------------|
   item 5
   item 6

new IntersectionObserver((entries) => {
  console.log(entries);
});

// Result: [
//   { id: 1, isIntersection: true, ... },
//   { id: 2, isIntersection: true, ... },
//   { id: 3, isIntersection: true, ... },
//   { id: 4, isIntersection: true, ... },
//   { id: 5, isIntersection: false, ... },
//   { id: 6, isIntersection: false, ... },
// ]
```

When scrolling to where `item 6` is visible and `item 1` isn't will result in this 
```
   item 1
|--------------------------------------------------|
|  item 2                                          |
|  item 3                                          |
|  item 4                                          |
|  item 5                                          |
|--------------------------------------------------|
   item 6


new IntersectionObserver((entries) => {
  console.log(entries);
});

// Result:

// Result: [
//   { id: 1, isIntersection: false, ... },
//   { id: 6, isIntersection: true, ... },
// ]
```

This will trigger item 6 to be `active` where actually item 2 is the first visible item.

This means that solely depending on the `IntersectionObserver` is not possible. Also when there is nog change in visible elements it can be possible that another scroll-spy target should be made active.

A possible correct implementation would be to combine the `IntersectionObserver` with a `onscroll` where the `IntersectionObserver` is used to save all visible elements in some array/object and the `onscroll` callback uses these visible elements to determine which one is the first visible (of course only intersecting elements can be active), I expect this first option will reduce the load of the `onscroll` callback. Or only using a `onscroll` (basically going back to version 5.1).

**Comment by qofe** [comment]
>
> > My issue was a conflict I had in the 'overflow' attribute on the 'body'. Once I cleared that up, it worked as expected.
> 
> @qofe Can you elaborate on this?

Yes, I had
`body {overflow-x:hidden; }
in top of my styles. Removed overflow setting, changed nothing else, and that eliminated the scrollspy issue described in the first post.

**Comment by davidtmiller** [comment]
I have been getting better results when wrapping the entire section that I want to spy on with the `id` target.

For example, instead of doing this (example given in docs):

```
<h4 id="scrollspyHeading1">First heading</h4>
<p>...</p>
```

I will do this:

```
<div id="scrollspyHeading1">
    <h4>First heading</h4>
    <p>...</p>
</div>
```

This way the `id` target is in the viewport for most of the time. This does seem to mess things up a bit when I have one section that is smaller than another that is also within the viewport.

My implementation is very similar to the old Bootstrap 3 docs with nested nav items that appear when the parent has the active state (like the sidebar here https://getbootstrap.com/docs/3.4/components/) making it very important to have the parent being spied on at all time.

I played around with the `rootMargin` property as well, similar to what was done in the docs, bringing it to 35% in my case which seemed to help.

Everything seems to be working when scrolling down, but when reaching the bottom of the page and scrolling back up, there are a lot of issues. The scroll position when clicking on menu items going back up is off, and some items don't receive the active state at all. I am thinking that this has to do with the last scrollspy section being very short in height, so it messes with the positioning when going back up.

Just some observations and suggestions in case they help progress this issue further!

**Comment by Fatbat** [comment]
> I have been getting better results when wrapping the entire section that I want to spy on with the `id` target.

Yeah, I wrap the entire section that references each scroll spy target and it doesn't help. The "Home" entry isn't working either. This has been broken for awhile now.

**Comment by iMattPro** [comment]
So this issue just being tossed aside?

**Comment by julien-deramond** [comment]
> So this issue just being tossed aside?

It's not being tossed aside. We're focusing right now on the v5.3.0 release. Then we'll try to define and prioritize what comes next. I'll keep in mind this one. Please understand that there's a huge amount of work to do in Bootstrap, and everybody does it in their spare time.

**Comment by itchyny** [comment]
In my use case, ScrollSpy does not work with sections with large height by default, and l noticed that the default threshold `[0.1, 0.5, 1]`, typically `0.1`, is not good for large height section . For me, `data-bs-threshold="0,1" data-bs-root-margin="-30% 0% -70%"` fixed the issue.

**Comment by gettonet** [comment]
> Is there any update on this? Scrollspy and the Docs are awful now. Before Scrollspy when spying on a list of links, would highlight when the target reached the top of the viewport, but now it highlights just when the target is visible in the window anywhere. There's absolutely no guidance on how to tweak these root margins and thresholds to at least recreate the original scroll spy behavior from before 5.2 where it highlights when the target is at the top of the viewport.

I completely agree and have the same issue. Following the topic...

**Comment by iMattPro** [comment]
> > So this issue just being tossed aside?
> 
> It's not being tossed aside. We're focusing right now on the v5.3.0 release. Then we'll try to define and prioritize what comes next. I'll keep in mind this one. Please understand that there's a huge amount of work to do in Bootstrap, and everybody does it in their spare time.

All I meant was, this Issue is no longer assigned to any project, so I see this falling through the cracks. It's already been an open issue and a problem for 18 months.

**Comment by M-Yankov** [comment]
Another example that scrollspy is not working in Bootstrap 5.3.2
Windows 10 64-bit Firefox 112.0,  (MS Edge 120.0.2210.144)
https://jsfiddle.net/2oxrkzyb/

If you swap resources to 5.1.3 it's working fine.

**Comment by nopenopenope** [comment]
What I figured is that the ScrollSpy seems to work when wrapping the seciont and making clear sections with that.

However, especially in responsive view (iPhone - roughly 400px) - it seems that very large sections are again not properly tracked by ScrollSpy. I am just checking out if it is an issue with the zIndex but that'd surprise me as of now. 

For example my whole page has a total height of 16630 px, and the largest section clocks in with 7350 pixels. I'd add screenshots but its not giving much information in responsive view - one time the active class is added to a nav, one time it isn't.

**Comment by PeterStavrou** [comment]
Has anyone managed to fix this? Even on the documentation page for v5.3 it's not working as it should.

**Comment by cwildfoerster** [comment]
any updates on this?

**Comment by joshvillarreal** [comment]
I am having similar issues!

**Comment by frontenddevguy** [comment]
I also had to wrap the section in a div with the target id, rather than rely on headings being correctly detected.
That improves the behaviour somewhat, and avoids there being 'dead space' between highlighted items when there are long chunks of text. 

However, the position of scroll where the id is detected seems erratic, and even depends on speed of scroll.

**Comment by frontenddevguy** [comment]
> > So this issue just being tossed aside?
> 
> It's not being tossed aside. We're focusing right now on the v5.3.0 release. Then we'll try to define and prioritize what comes next. I'll keep in mind this one. Please understand that there's a huge amount of work to do in Bootstrap, and everybody does it in their spare time.

Please just revert it back to the previous working version

**Comment by Fatbat** [comment]
> > So this issue just being tossed aside?
> 
> It's not being tossed aside. We're focusing right now on the v5.3.0 release. Then we'll try to define and prioritize what comes next. I'll keep in mind this one. Please understand that there's a huge amount of work to do in Bootstrap, and everybody does it in their spare time.

It has now been more than two years.

**Comment by DaHnDell** [comment]
Scrollspy is still not working, via BootStrap/5.3.3

**Comment by codyrobbins** [comment]
It’s insane that this has been completely broken for coming up on three years. As @frontenddevguy suggested, can the intersection observer “improvements” just be reverted to bring it back to its previously functioning state?

**Comment by jak-kal** [comment]
Definitely a regression - it does not work as intended. Not sure who passed this?

**Comment by razvanphp** [comment]
Rolling back to an old, performance-intensive code with scroll events is not the way to go! It's clear that browsers [natively](https://caniuse.com/intersectionobserver) support this use-case with [Intersection Observer API](https://medium.com/@mz.ebrahimi/discover-the-magic-of-intersection-observer-bringing-web-interactivity-to-life-716c742d02ac) and we should not go back and reinvent the wheel.

What I'm trying to understand is if this is a bug, change on default behaviour or a skill issue. I was out of bootstrap community for many years, but I must say, a lot of developers around here are just complaining and demanding free stuff. What if you would try to contribute back? BS is open-source. Fork it, add tests, fix it and push it back for everybody to benefit!

**Comment by kaldesigns** [comment]
> Rolling back to an old, performance-intensive code with scroll events is **not** the way to go! It's clear that browsers [natively](https://caniuse.com/intersectionobserver) support this use-case with [Intersection Observer API](https://medium.com/@mz.ebrahimi/discover-the-magic-of-intersection-observer-bringing-web-interactivity-to-life-716c742d02ac) and we should not go back and reinvent the wheel.
> 
> What I'm trying to understand is if this is a bug, change on default behaviour or a skill issue. I was out of bootstrap community for many years, but I must say, a lot of developers around here are just complaining and demanding free stuff. What if you would try to contribute back? BS is open-source. Fork it, add tests, fix it and push it back for everybody to benefit!

It's not a skill issue - it is quite clearly a regression in my opinion. That isn't to say we should roll back. Have you actually used it?

**Comment by codyrobbins** [comment]
> Rolling back to an old, performance-intensive code with scroll events is **not** the way to go!

It absolutely is the way to go if the old code worked and the new code is broken—having it work correctly is more important than it being performant.

> What I'm trying to understand is if this is a bug, change on default behaviour or a skill issue.

It’s a bug. I’m not sure how it adds to the conversation to suggest it’s a skill issue.

> I was out of bootstrap community for many years, but I must say, a lot of developers around here are just complaining and demanding free stuff. What if you would try to contribute back? BS is open-source. Fork it, add tests, fix it and push it back for everybody to benefit!

Everyone here is reporting and discussing a bug, not complaining and demanding free stuff. Telling people not to report bugs unless they fix it themselves is not productive. And there are, in fact, multiple people working on PRs to fix it.

**Comment by iMattPro** [comment]
Yeah, rolling back is not problematic. I've been stuck on 5.1.x for years now, just because Scrollspy is integral to the UX of my app, and having it work is more important than whatever was trying to be accomplished since Scrollspy broke. And you know, 5.1.x is working just fine - I just don't get to enjoy the native BS light/dark modes.

**Comment by JasonL-HealthTrust** [comment]
Just adding my voice as someone experiencing these problems with Scrollspy in Bootstrap 5.3.5. I used to use different (jQuery) methods to accomplish this and tried to switch to using Scrollspy with Boostrap 5.3.5. I could not understand why it was behaving so strangely until I saw this Issue, where everyone is seeing the unexpected behavior that I'm seeing.

**Comment by robsd** [comment]
> Yeah, rolling back is not problematic. I've been stuck on 5.1.x for years now, just because Scrollspy is integral to the UX of my app, and having it work is more important than whatever was trying to be accomplished since Scrollspy broke. And you know, 5.1.x is working just fine - I just don't get to enjoy the native BS light/dark modes.

It's annoying we have to do this, but I've just had to do the same for one of my client websites cos I was pulling my hair out trying to make it work by playing around with offset, root margin and custom CSS hacks

I'm guessing there are no decent 3rd party options that anyone has tried?

Bootstrap really need to sort this out

**Comment by guilhas07** [comment]
Could you test https://github.com/twbs/bootstrap/pull/41016 and see if it fixes your problems?

**Comment by DenisLopatin** [comment]
@razvanphp yes, and then you will dive into the component's work, find errors in testing, add your tests, fix the component, and your comit will be closed for the reason: sorry, a lot of code. Sorry, but how could there be less of it if the component is just broken from all sides, including tests that objectively don't work.t and there is 0 benefit from them

**Comment by mdo** [comment]
The ScrollSpy rewrite just landed on the `v6-dev` branch (#42557). It's now driven by an `IntersectionObserver` activation line (no scroll-position polling), with deterministic active-section detection, correct URL-hash/focus handling on smooth scroll, and `getElementById`-based id resolution (so dotted/encoded ids work) — which addresses this.

Please give the new ScrollSpy a try on `v6-dev` and share any feedback here. There's still time to refine it, both now and through the alpha, so if the behavior isn't right for your case we'd like to hear about it.

Closing as resolved by the rewrite — reopen or comment if it still reproduces on `v6-dev`.
